### ISSUE TITLE
inconsistent 422 errors

### ISSUE THREAD

**Issue by murtyk** [lyang/braintree-rails#29]
here is a script that reflects typical process in our workflow

```` ruby
def bt_process
  customer = Braintree::Customer.create(
    id:         1,
    first_name: 'John',
    last_name:  'Smith',
    email:      'john@example.com',
    payment_method_nonce: 'fake-valid-mastercard-nonce'
  )

  puts customer.errors.first if customer.try(:errors).try(:any?)

  bt_customer = BraintreeRails::Customer.find 1

  transaction = bt_customer.transactions.create(amount: 20)

  puts transaction.id
  
  bt2_customer = BraintreeRails::Customer.find 1

  bt2_customer.delete

  "complete"

rescue StandardError => error
  puts error
  "Not Complete" 
end
````

When I run this in rails console multiple times, the results are inconsistent:

[2] pry(main)> bt_process
I, [2015-11-15T07:46:48.351445 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:46:48 UTC] POST /customers 201
I, [2015-11-15T07:46:48.909273 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:46:48 UTC] GET /customers/1 200
I, [2015-11-15T07:46:49.983855 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:46:49 UTC] POST /transactions 201
I, [2015-11-15T07:46:50.535718 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:46:50 UTC] POST /transactions/advanced_search_ids 200
I, [2015-11-15T07:46:51.173253 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:46:51 UTC] POST /transactions/advanced_search 200
7kkcjt
I, [2015-11-15T07:46:51.822026 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:46:51 UTC] GET /customers/1 200
I, [2015-11-15T07:46:52.375280 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:46:52 UTC] DELETE /customers/1 200
=> "complete"
[3] pry(main)> bt_process
I, [2015-11-15T07:46:56.423873 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:46:56 UTC] POST /customers 201
I, [2015-11-15T07:46:57.115292 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:46:57 UTC] GET /customers/1 200
I, [2015-11-15T07:46:57.878768 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:46:57 UTC] POST /transactions 422
undefined method `for' for nil:NilClass
=> "Not Complete"
[4] pry(main)> bt_process
I, [2015-11-15T07:47:00.705432 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:00 UTC] POST /customers 422
#<Braintree::ValidationError:0x00000012a02028>
I, [2015-11-15T07:47:01.253328 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:01 UTC] GET /customers/1 200
I, [2015-11-15T07:47:02.178965 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:02 UTC] POST /transactions 422
undefined method `for' for nil:NilClass
=> "Not Complete"
[5] pry(main)> bt_process
I, [2015-11-15T07:47:04.018155 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:04 UTC] POST /customers 422
#<Braintree::ValidationError:0x0000000709b030>
I, [2015-11-15T07:47:04.625897 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:04 UTC] GET /customers/1 200
I, [2015-11-15T07:47:05.404610 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:05 UTC] POST /transactions 422
undefined method `for' for nil:NilClass
=> "Not Complete"
[6] pry(main)> bt_process
I, [2015-11-15T07:47:07.445380 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:07 UTC] POST /customers 422
#<Braintree::ValidationError:0x00000009d5d3e8>
I, [2015-11-15T07:47:08.022609 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:08 UTC] GET /customers/1 200
I, [2015-11-15T07:47:08.765681 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:08 UTC] POST /transactions 422
undefined method `for' for nil:NilClass
=> "Not Complete"
[7] pry(main)> bt_process
I, [2015-11-15T07:47:10.475470 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:10 UTC] POST /customers 422
#<Braintree::ValidationError:0x0000000a90cf18>
I, [2015-11-15T07:47:11.027664 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:11 UTC] GET /customers/1 200
I, [2015-11-15T07:47:11.855194 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:11 UTC] POST /transactions 422
undefined method `for' for nil:NilClass
=> "Not Complete"
[8] pry(main)> bt_process
I, [2015-11-15T07:47:13.459150 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:13 UTC] POST /customers 422
#<Braintree::ValidationError:0x0000000afb4050>
I, [2015-11-15T07:47:13.996188 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:13 UTC] GET /customers/1 200
I, [2015-11-15T07:47:14.798545 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:14 UTC] POST /transactions 422
undefined method `for' for nil:NilClass
=> "Not Complete"
[9] pry(main)> bt_process
I, [2015-11-15T07:47:16.805844 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:16 UTC] POST /customers 422
#<Braintree::ValidationError:0x0000000b750548>
I, [2015-11-15T07:47:17.332318 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:17 UTC] GET /customers/1 200
I, [2015-11-15T07:47:18.096582 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:18 UTC] POST /transactions 422
undefined method `for' for nil:NilClass
=> "Not Complete"
[10] pry(main)> bt_process
I, [2015-11-15T07:47:19.867987 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:19 UTC] POST /customers 422
#<Braintree::ValidationError:0x0000000bf95aa0>
I, [2015-11-15T07:47:20.410591 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:20 UTC] GET /customers/1 200
I, [2015-11-15T07:47:21.155223 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:21 UTC] POST /transactions 201
I, [2015-11-15T07:47:21.675826 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:21 UTC] POST /transactions/advanced_search_ids 200
I, [2015-11-15T07:47:22.326993 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:22 UTC] POST /transactions/advanced_search 200
68wcy8
I, [2015-11-15T07:47:22.968598 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:22 UTC] GET /customers/1 200
I, [2015-11-15T07:47:23.552406 #25108]  INFO -- : [Braintree] [15/Nov/2015 12:47:23 UTC] DELETE /customers/1 200
=> "complete"
[11] pry(main)>

**Comment by lyang** [comment]
Hi @murtyk , thank you for reporting the issue.

To help diagnose the issue, would you mind set the log level to debug, so that the request/response xml would be logged? For e.g.
```ruby
logger = Logger.new('path/to/braintree.log')
logger.level = Logger::DEBUG
BraintreeRails::Configuration.logger = logger
```
It also appears that you are testing in the sandbox, correct?

**Comment by lyang** [comment]
Hi @murtyk,
I think you have triggered the [duplicate check](https://articles.braintreepayments.com/control-panel/transactions/duplicate-checking) in the Braintree gateway, where it saw identical transaction requests within 30 seconds, so the subsequent ones were gateway rejected.

I will look into fix the `braintree-rails` gem so that it handles this properly. But I think to fix the underlying root cause of the `422` errors, you'll have to randomize your test data. For [example](https://github.com/lyang/braintree-rails/blob/master/spec/support/helper.rb#L51-L56)

Let me know if this helps.

**Comment by murtyk** [comment]
Hi @lyang,

In the spec, we are creating and destroying users for different contexts and user.id is same (= 1) in all cases. I guess this might be causing duplicates since we use user.id for customer.id. This should not happen in production. We will anyway explore in generating unique customer id for each user.

Do you still need the debug logs?

**Comment by lyang** [comment]
According to the support article, those are the [params](https://articles.braintreepayments.com/control-panel/transactions/duplicate-checking#apple-pay-and-credit-cards) you can randomize to workaround the duplicate check.

I would suggest turn off duplicate check *only in sandbox* as the last resort. If you do still rely on duplicate checking in some scenarios, you can create a separate sandbox account for that.

Don't worry about the log, I can reproduce it locally. But it would still be good if you can confirm it's indeed the duplicate check issue on your side.

**Comment by lyang** [comment]
Hi @murtyk ,
This issue should be fixed now in `1.4.1`, let me know if you still see it blowing up.
